### PR TITLE
Give RawBytes an error-checking smart constructor

### PR DESCRIPTION
**Pull request by dcoutts**
It checks the byte array slice is valid.

And make sure the constructor is used in various places, which means not using coerce so much.

### CODE REVIEW

File: src/Database/LSMTree/Internal/PageAcc.hs [hunk 1]
```diff
@@ -317,8 +317,10 @@ serializePageAccN size MPageAcc {..} = do
             | otherwise = do
                   -- traceM $ "loop " ++ show (i, ko, kd, vo, vd)
 
-                  SerialisedKey   (RawBytes (PV.Vector koff klen kba)) <- MV.read mpKeys i
-                  SerialisedValue (RawBytes (PV.Vector voff vlen vba)) <- MV.read mpValues i
+                  k <- MV.read mpKeys i
+                  v <- MV.read mpValues i
+                  let SerialisedKey   (RawBytes (PV.Vector koff klen kba)) = k
```
Comment (phadej) [inline src/Database/LSMTree/Internal/PageAcc.hs:322]: why this change?

Is having lazy pattern match important?

File: src/Database/LSMTree/Internal/Serialise/RawBytes.hs [hunk 2]
```diff
@@ -76,9 +78,25 @@ import           GHC.Word
...
+    RawBytes v = mkRawBytes v
+
+mkRawBytes :: P.Vector Word8 -> RawBytes
+mkRawBytes (P.Vector off len ba) =
+    assert (isValidSlice off len ba) $
+    RawBytes (P.Vector off len ba)
+
+isValidSlice :: Int -> Int -> ByteArray -> Bool
+isValidSlice off len ba =
+    len >= 0
```
Comment (phadej) [inline src/Database/LSMTree/Internal/Serialise/RawBytes.hs:97]: Add `off >=0`  too?

**Review by phadej — COMMENTED** [review]
I feel that what we need is a smart constructor for `P.Vector`, not for `RawBytes`.

It's the vectors invariants we break, not the `RawBytes` (which doesn't have any extra).

File: src/Database/LSMTree/Internal/Serialise/RawBytes.hs [hunk 3]
```diff
@@ -211,10 +229,10 @@ fromByteArray :: Int -> Int -> ByteArray -> RawBytes
 fromByteArray off len ba = RawBytes (P.Vector off len ba)
 
 pack :: [Word8] -> RawBytes
-pack = coerce P.fromList
+pack = RawBytes . P.fromList
```
Comment (phadej) [inline src/Database/LSMTree/Internal/Serialise/RawBytes.hs:232]: e.g. here we "unnecessarily" double-check that `P.fromList` doesn't violate `Vector`s invariants. This doesn't feel right to be so overly paranoid.

File: src/Database/LSMTree/Internal/PageAcc.hs (hunk 1 above)
Comment (dcoutts) [reply src/Database/LSMTree/Internal/PageAcc.hs:322]: Unfortunately having a pattern synonym means the pattern match is always considered partial for the purpose of `MonadFail`. So if one uses the pattern here in the monad bind then we get complaints about not having a fail constraint. Using COMPLETE pragmas doesn't help apparently.

We can make those pattern matches strict.

**Comment by dcoutts** [comment]
> I feel that what we need is a smart constructor for `P.Vector`, not for `RawBytes`.
> 
> It's the vectors invariants we break, not the `RawBytes` (which doesn't have any extra).

True, but we're defining `RawBytes`. There's no so much we can enforce about `P.Vector`. Any concrete suggestion?

File: src/Database/LSMTree/Internal/PageAcc.hs (hunk 1 above)
Comment (phadej) [reply src/Database/LSMTree/Internal/PageAcc.hs:322]: Ah, so this https://gitlab.haskell.org/ghc/ghc/-/issues/15681

I kindly ask you to mention that you are hitting that as well. Maybe at some point, GHC HQ acknowledges that that issue is really should be fixed even if you cannot get ICFP paper out of that.

**Comment by phadej** [comment]
I'd make a `Data.LSMTree.Internal.Vector.Primitive` module re-exporting most of `Data.Vector.Primitive`, but replacing the constructor.

But I really think that we should just use smart constructor function, as MonadFail and pattern synonym issue is annoying.  Pattern synonyms are "broken", and until they are fixed I'd avoid their usage.

File: src/Database/LSMTree/Internal/Serialise/RawBytes.hs [hunk 4]
```diff
@@ -76,9 +78,25 @@ import           GHC.Word
...
+{-# COMPLETE RawBytes #-}
+pattern RawBytes :: P.Vector Word8 -> RawBytes
+pattern RawBytes v <- MkRawBytes v
+  where
+    RawBytes v = mkRawBytes v
+
+mkRawBytes :: P.Vector Word8 -> RawBytes
+mkRawBytes (P.Vector off len ba) =
+    assert (isValidSlice off len ba) $
+    RawBytes (P.Vector off len ba)
```
Comment (phadej) [inline src/Database/LSMTree/Internal/Serialise/RawBytes.hs:93]: This loops.

Comment (dcoutts) [reply src/Database/LSMTree/Internal/Serialise/RawBytes.hs:93]: :facepalm:

**Comment by dcoutts** [comment]
> I'd make a `Data.LSMTree.Internal.Vector.Primitive` module re-exporting most of `Data.Vector.Primitive`, but replacing the constructor.
> 
> But I really think that we should just use smart constructor function, as MonadFail and pattern synonym issue is annoying. Pattern synonyms are "broken", and until they are fixed I'd avoid their usage.
> 
> (It shouldn't be "hard" to fix, if you can bind and then `let`, and there is no incomplete pattern match warning, GHC is clearly smart enough to figure out COMPLETEness.)

I'd be happy to change `RawBytes` to be a more complete abstraction rather than exposing it as being a `P.Vector`.

This is the thing that users are going to have to use as the source/dest of their (de)serialisation. So probably the thing to do is just provide a bunch of useful conversion functions (with appropriate invariant checking).

I could retire this PR and make one that makes it properly abstract.
